master/internal/db/postgres_trial_metrics.go

```diff
@@ -81,26 +138,97 @@ WHERE trial_id = $1
 	return int(affectedRows), nil
 }
 
-func (db *PgDB) addRawMetrics(ctx context.Context, tx *sqlx.Tx, metricsBody *map[string]interface{},
+// addMetricsWithMerge inserts a set of metrics to the database and returns the metric id,
+// and the results.
+func (db *PgDB) addMetricsWithMerge(ctx context.Context, tx *sqlx.Tx, mBody *metricsBody,
+	runID, trialID, lastProcessedBatch int32, mType model.MetricType,
+) (metricID int, addedMetrics *metricsBody, err error) {
+	var existingBodyJSON model.JSONObj
+	err = tx.QueryRowContext(ctx, `
+SELECT COALESCE((SELECT metrics FROM metrics
+WHERE archived = false
+AND total_batches = $1
+AND trial_id = $2
+AND partition_type = $3
+AND custom_type = $4), NULL)`,
+		lastProcessedBatch, trialID,
+		customMetricTypeToPartitionType(mType), mType).Scan(&existingBodyJSON)
+	if err != nil {
+		return 0, nil, errors.Wrap(err, "getting old metrics")
+	}
+	needsMerge := existingBodyJSON != nil
+
+	fmt.Println("existing body", existingBodyJSON)
+	fmt.Println("needs merge", needsMerge)
+
+	if !needsMerge {
+		id, err := db.addRawMetrics(ctx, tx, mBody, runID, trialID, lastProcessedBatch, mType)
+		return id, mBody, err
+	}
+
+	existingBody := &metricsBody{Type: mType}
+	// CHECK: how do we avoid this copy?
+	if err = existingBody.LoadJSON(&existingBodyJSON); err != nil {
+		return 0, nil, err
+	}
+	fmt.Println("existing body avg", existingBody.AvgMetrics)
+	finalBody, err := shallowUnionMetrics(existingBody, mBody)
+	if err != nil {
+		return 0, nil, err
+	}
+	id, err := db.updateRawMetrics(ctx, tx, finalBody, runID, trialID, lastProcessedBatch, mType)
+	return id, mBody, err
+}
+
+func (db *PgDB) updateRawMetrics(ctx context.Context, tx *sqlx.Tx, mBody *metricsBody,
 	runID, trialID, lastProcessedBatch int32, mType model.MetricType,
 ) (int, error) {
+	if err := mType.Validate(); err != nil {
+		return 0, err
+	}
 	pType := customMetricTypeToPartitionType(mType)
+	fmt.Println("updating metrics", mBody)
+
+	var metricRowID int
+	if err := tx.QueryRowContext(ctx, `
+UPDATE metrics
+SET metrics = $1
+WHERE archived = false
+AND trial_id = $2
+AND partition_type = $3
+AND custom_type = $4
+AND total_batches = $5
+RETURNING id`,
+		*mBody.ToJSONObj(), trialID, pType, mType, lastProcessedBatch,
+	).Scan(&metricRowID); err != nil {
+		return 0, errors.Wrap(err, "updating metrics")
+	}
+
+	return metricRowID, nil
+}
 
+// addRawMetrics inserts a set of raw metrics to the database and returns the metric id.
+func (db *PgDB) addRawMetrics(ctx context.Context, tx *sqlx.Tx, mBody *metricsBody,
+	runID, trialID, lastProcessedBatch int32, mType model.MetricType,
+) (int, error) {
 	if err := mType.Validate(); err != nil {
 		return 0, err
 	}
+	pType := customMetricTypeToPartitionType(mType)
 
 	var metricRowID int
+	// ON CONFLICT clause is not supported with partitioned tables (SQLSTATE 0A000)
```

Comment:
good to know.

it'd probably also be awkward to implement "merge jsons, but only if top level keys don't intersect" and return a sensible error message about this from `ON CONFLICT`.

Reply:
it was before realizing the issue with conflicting keys on min/max. also I've been trying to minimize added friction to the code path for adding metrics that wouldn't need merging but at the end I couldn't avoid the lookup.
get-modify-update vs update-fail-get-modify-update or do parts of this in the db